service/controller/resource/clusterconfigmap/desired.go

```diff
@@ -67,56 +68,55 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) ([]*cor
 		clusterCA = string(secret.Data["tls.crt"])
 	}
 
-	appOperatorValues := map[string]interface{}{
-		"app": map[string]interface{}{
-			"watchNamespace":    cr.GetNamespace(),
-			"workloadClusterID": key.ClusterID(&cr),
-		},
-		"provider": map[string]interface{}{
-			"kind": r.provider,
-		},
-		"registry": map[string]interface{}{
-			"domain": r.registryDomain,
-		},
-	}
+	// By default, the IP for the coredns Service belongs to the pod CIDR,
+	// but if "serviceSubnet" is set in KubeadmControlPlane CR then we want to override it with an IP belonging
+	// to the service CIDR.
+	var clusterDNSIP = r.dnsIP
+	{
+		if cr.Spec.ControlPlaneRef != nil {
+			kubeadmControlPlane := &unstructured.Unstructured{}
+			kubeadmControlPlane.SetGroupVersionKind(schema.GroupVersionKind{
+				Group:   cr.Spec.ControlPlaneRef.GroupVersionKind().Group,
+				Kind:    cr.Spec.ControlPlaneRef.Kind,
+				Version: cr.Spec.ControlPlaneRef.GroupVersionKind().Version,
+			})
+			err = r.k8sClient.CtrlClient().Get(ctx, client.ObjectKey{
+				Namespace: cr.Namespace,
+				Name:      cr.Spec.ControlPlaneRef.Name,
+			}, kubeadmControlPlane)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
 
-	clusterValues := map[string]interface{}{
-		"baseDomain": key.BaseDomain(&cr, r.baseDomain),
-		"chartOperator": map[string]interface{}{
-			"cni": map[string]interface{}{
-				"install": true,
-			},
-		},
-		"cluster": map[string]interface{}{
-			"calico": map[string]interface{}{
-				"CIDR": podCIDR,
-			},
-			"kubernetes": map[string]interface{}{
-				"API": map[string]interface{}{
-					"clusterIPRange": r.clusterIPRange,
-				},
-				"DNS": map[string]interface{}{
-					"IP": r.dnsIP,
-				},
-			},
-		},
-		"clusterCA":    clusterCA,
-		"clusterDNSIP": r.dnsIP,
-		"clusterID":    key.ClusterID(&cr),
-		"clusterCIDR":  "",
-		"provider":     "unknown",
+			serviceCidr, serviceCidrFound, err := unstructured.NestedString(kubeadmControlPlane.Object, []string{"spec", "kubeadmConfigSpec", "clusterConfiguration", "networking", "serviceSubnet"}...)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+
+			if !serviceCidrFound {
+				serviceCidr = DefaultK8sServiceCidr
+			}
+
+			clusterDNSIP, err = key.DNSIP(serviceCidr)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+		}
 	}
 
+	var (
+		// clusterCIDR is are only used on azure.
+		clusterCIDR = ""
+		// gcpProject is only used on gcp.
+		gcpProject = ""
+		// region is only used on gcp.
+		region = ""
```

Comment:
I would prefer to name it gcpRegion as one day there might be an awsRegion as well.
As long as you don't say the same var can be used on both providers

Reply:
I removed the region as we don't have an specific use case for it right now.